R/merge.R

```diff
@@ -96,8 +96,19 @@ merge.data.table = function(x, y, by = NULL, by.x = NULL, by.y = NULL, all = FAL
   if (all.y && nrow(y)) {  # If y does not have any rows, no need to proceed
     # Perhaps not very commonly used, so not a huge deal that the join is redone here.
     missingyidx = y[!x, which=TRUE, on=by, allow.cartesian=allow.cartesian]
+    # TO DO: replace by following once #5446 is merged
+    # if (length(missingyidx)) dt = rbind(dt, y[missingyidx], use.names=FALSE, fill=TRUE, ignore.attr=TRUE)
     if (length(missingyidx)) {
-      dt = rbind(dt, y[missingyidx], use.names=FALSE, fill=TRUE)
+      yy = y[missingyidx]
+      othercolsx = setdiff(nm_x, by)
+      if (length(othercolsx)) {
+        tmp = rep.int(NA_integer_, length(missingyidx))
+        # TO DO: use set() here instead..
+        yy = cbind(yy, x[tmp, othercolsx, with = FALSE])
```

Comment:
I think this could use a comment, the construction looks a bit unusual to me. We do this to get a rectangle of `NA`s of the correct shape _and_ data types, is that right?

Reply:
Regarding comnent. This probably could be nicely implemented with mergelist and cbindlist which can avoid copies via copy arg.

Reply:
(resolving as my original comment was addressed, feel free to move TODOs re:cbindlist/mergelist to follow-up issues)